python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -1302,64 +1313,115 @@ def _ensure_persisted_execution_plan_snapshot(
 
         return execution_plan_snapshot_id
 
-    def _log_asset_planned_events(
-        self, dagster_run: DagsterRun, execution_plan_snapshot: "ExecutionPlanSnapshot"
+    def _log_materialization_planned_event_for_asset(
+        self,
+        dagster_run: DagsterRun,
+        asset_key: AssetKey,
+        job_name: str,
+        step: "ExecutionStepSnap",
+        output: "ExecutionStepOutputSnap",
+        job_partitions_def: Optional["PartitionsDefinition"],
     ) -> None:
+        from dagster._core.definitions.partition import DynamicPartitionsDefinition
+        from dagster._core.definitions.partition_key_range import PartitionKeyRange
         from dagster._core.events import (
             AssetMaterializationPlannedData,
             DagsterEvent,
             DagsterEventType,
         )
 
+        partition_tag = dagster_run.tags.get(PARTITION_NAME_TAG)
+        partition_range_start, partition_range_end = (
+            dagster_run.tags.get(ASSET_PARTITION_RANGE_START_TAG),
+            dagster_run.tags.get(ASSET_PARTITION_RANGE_END_TAG),
+        )
+
+        if partition_tag and (partition_range_start or partition_range_end):
+            raise DagsterInvariantViolationError(
+                f"Cannot have {ASSET_PARTITION_RANGE_START_TAG} or"
+                f" {ASSET_PARTITION_RANGE_END_TAG} set along with"
+                f" {PARTITION_NAME_TAG}"
+            )
+
+        partitions_subset = None
+        if partition_range_start or partition_range_end:
+            if not partition_range_start or not partition_range_end:
+                raise DagsterInvariantViolationError(
+                    f"Cannot have {ASSET_PARTITION_RANGE_START_TAG} or"
+                    f" {ASSET_PARTITION_RANGE_END_TAG} set without the other"
+                )
+
+            if job_partitions_def is None:
+                raise DagsterInvariantViolationError(
+                    "Must provide partitions_def to create_run when creating "
+                    "a run with a partition range."
+                )
+
+            if (
+                isinstance(job_partitions_def, DynamicPartitionsDefinition)
+                and job_partitions_def.name is None
+            ):
+                raise DagsterInvariantViolationError(
+                    "Creating a run targeting a partition range is not supported for jobs partitioned with function-based dynamic partitions"
+                )
+
+            # For now, yielding materialization planned events for single run backfills
+            # is only supported on cloud
+            if self.is_cloud_instance and check.not_none(output.properties).is_asset_partitioned:
```

Comment:
Is the flag necessary?  Could we always populate the subset but just update the partitions table in the storage layer?

We subclass DagsterInstance in cloud, and use it as a place to override default behavior.  Adding an `is_cloud_instance` flag in OSS feels off to me and could promote more confusing intertwining of logic.

Reply:
This is a good call, I believe OSS functions the same regardless of whether `partition_key=None` vs `partitions_subset` is populated